README

Updated for version 3.10.0

This is the OPeNDAP netCDF Data Handler. It is used along with Hyrax.

For information about building the OPeNDAP netCDF Data Handler, see the
INSTALL file.

This data handler is a component of the OPeNDAP DAP Server; the server base
software is designed to allow any number of handlers to be configured easily.
See the DAP Server README and INSTALL files for information about
configuration, including how to use this handler.

This version of the handler has been updated to read netcdf 4 files
and supports most of the netcdf 4 data types, not just those defined
as part of the netcdf 'classic' data model.

See http://docs.opendap.org/index.php/BES_-_Modules_-_The_NetCDF_Handler
for the full story on the different data models and how they are
mapped into DAP2 along with the (expanded) set of configuration
parameters.

Test data are also installed, so after installing this handler, Hyrax 
will have data to serve providing an easy way to test your new installation
and to see how a working bes.conf should look. To use this, make sure
that you first install the bes, and that dap-server gets installed too.
Finally, every time you install or reinstall handlers, make sure to restart
the BES and OLFS.



NEWS
News for version 3.11.4

Updates since 2016-01-26, see the ChangeLog

News for version 3.11.3

Minor updates, see the ChangeLog

News for version 3.11.2

Minor updates, see the ChangeLog

News for version 3.11.1

Minor updates, see the ChangeLog

Version 3.10.4 (23 April 2014)

Changed default behavior for handelr so that it is explicitly set
to ShowSharedDimensions=true

Version 3.10.3 (10 September 2013)

Minor organizational changes.

Version 3.10.2

Merged to trunk; tagged.

Version 3.10.1

Fixes for netcdf 4.1.1, which is the version in RHEL6's RPM
package. These include _not_ using NC_FIRSTUSERTYPEID and the old
nctypelen(). Those work with netcdf 4.1.2, but not 4.1.1 and they are
probably not that gereat an idea anyway.

Version 3.10.0

This version introduces some major improvements to the handler. 
Now the handler reads netcdf 4 files and externalizes all but 
a very small number of its native data types (64-bit ints, vlens 
and groups are yet to be supported). This means that most netcdf 
4 files work, not just the 'classic netcdf' files.

Also added to the handler are several configuration parameters. Briefly, they are:
* IgnoreUnknowTypes
* ShowSharedDimensions
* PromoteByteToShort

See http://docs.opendap.org/index.php/BES_-_Modules_-_The_NetCDF_Handler for the full story.

Version 3.9.3

Build fixes.

Version 3.9.2 (14 Sept 2010)

The default option is now (re)set to include those extra map/Arrays.

Version 3.9.1 (5 May 2010)

The handler now implements its own transfer_attributes code to merge
attributes from its DAS response into the DDS.

The handler now builds with netcdf 4.1

Version 3.9.0 (3 Feb 2010)

Removed dejagnu tests and added autotool tests.
RPM to use nc.conf instead of bes-nc-data.sh.
API changes to BESKeys. 
Added some Shrew build changes for the shrew hyrax project. 
Minor libdap changes.
Removed server 3 code.
Added comd file that gets the DataDDX.
Made a few DAP macro changes.
Added info about the shared dimension special option.

Version 3.8.3 (9 June 2009)

Fixes the dejagnu tests.

Version 3.8.2 (16 March 2009)

Constraint fixes and support for the service responses.

version 3.8.1 (5 Feb 2009)

Fixed a problem with the new code to represent arrays of NC_CHAR as DAP
String variables.

version 3.8.0 (29 Jan 2009)

Cleaned up error handling so that wehn installed as a BES handler there's no
spurious output to stderr.

The handler no longer packages NC_CHAR arrays as arrays of DAP Strings - an
array of char is now a string and a rank N array of char is a rank N-1 array
of string. The DAP String variables have an new attribute called
'string_length' which gives a hint to code that needs to know how much memory
allocate for the string. Inthe rank N case the 'string_length' is te size of
the Nth dimension.

The code has been updated to work with the new libdap and bes.

The handler now has standalone tests using the besstandalone test driver.

Version 3.7.9 (04 Mar 2008)

pkg and dmg build fixes using PackageMaker and DropDMG. General
configuration upgrades.

libdap namespace addition

BES exception upgrade

Version 3.7.8 (03 Dec 2007)

pkg and dmg build fixes

Version 3.7.7 (14 Nov 2007)

Support for pkg-config added.

Don't edit bes.conf during make install, but document the bes-nc-dap.sh
script and the bes-conf make target in README.

Various bug fixes.

Version 3.7.6 (13 March 2007)

Default data and bes.conf with make install, same as the other handlers in
this release.

Version 3.7.5

Better error handling.

Build fixes.

Version 3.7.4

Updates for the 1.4 release of Server 4. Also included are bug fixes, memory
leak fixes, and performance enhencements.

Version 3.7.3

The handler now builds DDS objects that contain attributes and will thus work
with server-side functions that use attributes.

Added support for compressed data.

Added new module build options. Now modules use the libtool -module and 
-no-version options.

The module name is now set by the bes configuration file, not in this code.

Version 3.7.2

Fix for ticket 488. Embedded double quotes in string attributes were not
escaped. The server was modified so that String attributes are now passed
through the libdap function escattr(). There may be differences for some
String attributes in some clients as a result. Note that this version of the
handler requires libdap 3.7.0 or greater.

Version 3.7.1

This version includes a fix for bug 435. There are serveral data handlers
that use the same names for externally visible functions. That's fine when
the handlers are executables, but not so when they are shared object
libraries. The function named have been changed.

Version 3.7.0

This version of the netCDF handler includes an executable for use with
Server3 and a shared-object library (aka module) for use with Server4. This
is a source release to support the Server4 1.2 beta release and has not been
tested with Server3, although there should be no problems.

Version 3.6.0

This version of the netcdf handler builds with libdap 3.6.0.

Patrice Dumas has contributed improvements to the spec file used to build the
RPM packages.

Misc. fixes for the handler when used with Server4 (which is in development).

Version 3.5.2

Fix for ticket 281: int variable break on 64 bit machines. The fix involved
using nc_var_get_int() in place of nc_var_get_long in NCArray::read() and
using NC_INT in place of NC_LONG in read's switch statement.

Version 3.5.1

Bug fix: The handler did not process arrays of bytes correctly.

--------------------------------------------------------------------------

The handler now works with the new dap server configuration.

Removed the libnc-dods software. The OPeNDAP-netCDF client library is now
in a separate CVS module named libnc-dap. 

This software now uses automake and a modern autoconf. 

---------------------------------------------------------------------------

01/27/03 jhrg

This code now builds with gcc 3.2 in addition to VC++ and gcc 2.95.3.


NCUrl.cc

// -*- mode: c++; c-basic-offset:4 -*-

// This file is part of nc_handler, a data handler for the OPeNDAP data
// server. 

// Copyright (c) 2002,2003 OPeNDAP, Inc.
// Author: James Gallagher <jgallagher@opendap.org>
//
// This is free software; you can redistribute it and/or modify it under the
// terms of the GNU Lesser General Public License as published by the Free
// Software Foundation; either version 2.1 of the License, or (at your
// option) any later version.
// 
// This software is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
// License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// You can contact OPeNDAP, Inc. at PO Box 112, Saunderstown, RI. 02874-0112.
 

// (c) COPYRIGHT URI/MIT 1994-1996
// Please read the full copyright statement in the file COPYRIGHT.
//
// Authors:
//      reza            Reza Nekovei (reza@intcomm.net)

// netCDF sub-class implementation for NCByte,...NCGrid.
// The files are patterned after the subcalssing examples 
// Test<type>.c,h files.
//
// ReZa 1/12/95

#include "config_nc.h"

static char rcsid[] not_used ={"$Id$"};

#include <InternalErr.h>
#include "NCUrl.h"

NCUrl::NCUrl(const string &n, const string &d) : Url(n, d)
{
}

NCUrl::NCUrl(const NCUrl &rhs) : Url(rhs)
{
}

NCUrl::~NCUrl()
{
}

NCUrl &
NCUrl::operator=(const NCUrl &rhs)
{
    if (this == &rhs)
        return *this;

    dynamic_cast<NCUrl&>(*this) = rhs;

    return *this;
}


BaseType *
NCUrl::ptr_duplicate()
{
    return new NCUrl(*this);
}

// $Log: NCUrl.cc,v $
// Revision 1.11  2005/04/19 23:16:18  jimg
// Removed client side parts; the client library is now in libnc-dap.
//
// Revision 1.10  2005/04/08 17:08:47  jimg
// Removed old 'virtual ctor' functions which have now been replaced by the
// factory class code in libdap++.
//
// Revision 1.9  2005/03/31 00:04:51  jimg
// Modified to use the factory class in libdap++ 3.5.
//
// Revision 1.8  2005/02/17 23:44:13  jimg
// Modifications for processing of command line projections combined
// with the limit stuff and projection info passed in from the API. I also
// consolodated some of the code by moving d_source from various
// classes to NCAccess. This may it so that DODvario() could be simplified
// as could build_constraint() and store_projection() in NCArray.
//
// Revision 1.7  2005/01/26 23:25:51  jimg
// Implemented a fix for Sequence access by row number when talking to a
// 3.4 or earlier server (which contains a bug in is_end_of_rows()).
//
// Revision 1.6  2004/09/08 22:08:22  jimg
// More Massive changes: Code moved from the files that clone the netCDF
// function calls into NCConnect, NCAccess or nc_util.cc. Much of the
// translation functions are now methods. The netCDF type classes now
// inherit from NCAccess in addition to the DAP type classes.
//
// Revision 1.5  2004/03/08 19:13:06  jimg
// Removed unnecessary read() method implementation.
//
// Revision 1.4  2003/12/08 18:06:37  edavis
// Merge release-3-4 into trunk
//
// Revision 1.3  2003/09/25 23:09:36  jimg
// Meerged from 3.4.1.
//
// Revision 1.2.8.1  2003/09/06 23:33:14  jimg
// I modified the read() method implementations so that they test the new
// in_selection property. If it is true, the methods will read values
// even if the send_p property is not true. This is so that variables used
// in the selection part of the CE, or as function arguments, will be read.
// See bug 657.
//
// Revision 1.2  2000/10/06 01:22:03  jimg
// Moved the CVS Log entries to the ends of files.
// Modified the read() methods to match the new definition in the dap library.
// Added exception handlers in various places to catch exceptions thrown
// by the dap library.
//
// Revision 1.1  1999/07/28 00:22:45  jimg
// Added
//
// Revision 1.7  1999/05/07 23:45:33  jimg
// String --> string fixes
//
// Revision 1.6  1998/08/06 16:33:26  jimg
// Fixed misuse of the read(...) member function. Return true if more data
// is to be read, false is if not and error if an error is detected
//
// Revision 1.5  1996/09/17 17:06:45  jimg
// Merge the release-2-0 tagged files (which were off on a branch) back into
// the trunk revision.
//
// Revision 1.4.4.3  1996/09/17 00:26:33  jimg
// Merged changes from a side branch which contained various changes from
// Reza and Charles.
// Removed ncdump and netexec since ncdump is now in its own directory and
// netexec is no longer used.
//
// Revision 1.4.4.2  1996/07/10 21:44:28  jimg
// Changes for version 2.06. These fixed lingering problems from the migration
// from version 1.x to version 2.x.
// Removed some (but not all) warning generated with gcc's -Wall option.
//
// Revision 1.4.4.1  1996/06/25 22:04:48  jimg
// Version 2.0 from Reza.
//
// Revision 1.4  1995/07/09  21:33:53  jimg
// Added copyright notice.
//
// Revision 1.3  1995/06/28  20:22:41  jimg
// Replaced malloc calls with calls to new (and calls to free with calls to
// delete).
//
// Revision 1.2  1995/03/16  16:56:48  reza
// Updated for the new DAP. All the read_val mfunc. and their memory management
// are now moved to their parent class.
// Data transfers are now in binary while DAS and DDS are still sent in ASCII.
//
// Revision 1.1  1995/02/10  04:57:58  reza
// Added read and read_val functions.
